### ISSUE TITLE
Best practice of kubernetes cluster setup to support large number of services

### ISSUE THREAD

**Issue by nkwangleiGIT** [GoogleCloudPlatform/kubernetes#9673]
Hi
We're planning to use k8s as the framework as our cloud service and need to support a large number of services(pods). I'm not sure if we have any guide or best practices to setup such kind of environment, from architectural view, some related questions below:
1. Do we need to setup cluster for each k8s components? like api-server, etcd, scheduler, controller-manager
2. Ho we can achieve best performance if want to use kube-prorxy?
3. Any performance tuning parameters we can use to have best performance?

Any other suggestions will be appreciated, thanks.

**Comment by erictune** [comment]
How large is large?  We have tested with 100 nodes and 30 pods per node (3000 pods).  If your cloud service is much larger than that, then we should discuss further.

Are you planning to run on hosted VMs, or on-prem VMs, or bare metal?   That will help me help you pick one of the getting-started guides that will work for you, which is a prerequisite for answering a lot of your questions.

You asked "do we need to setup cluster for each k8s components? like api-server, etcd, scheduler, controller-manager".  When you use one of the getting-started-guides, it takes care of setting up a kubernetes cluster that includes those components.  Each of the components you mentioned (api-server, etcd, scheduler, controller-manager) is part of the same cluster.  If you are asking whether each of those components needs to be replicated for availability, we don't currently do that by default right now, but it is something we plan to offer in the future.

You also asked "How we can achieve best performance if want to use kube-prorxy".  You really need to use kube-proxy to take full advantage of kubernetes.   

Happy to discuss further.

**Comment by nkwangleiGIT** [comment]
Hi erictune, thanks for your comments.
1. We're doing some public container as a service, so we expect it'll be much larger, maybe one hundred thousand. If that's the scope, I think we need to have multiple clusters to support it, please also comment here.

2. We're running on other cloud provider(IaaS), maybe AWS, GCE, etc...

3. Yes, we already have the cluster running following the k8s guide, so I'm interested in
```
whether each of those components needs to be replicated for availability, we don't currently do that by default right now, but it is something we plan to offer in the future.
```
So for now, it's fine to have one service for each component?

4. About kube-proxy, we have thousands of rules created in iptables on each node that has kube-proxy runing, and we found that it's some slow sometimes to create new the iptable rule when create a new service, and got service unavailable issue when tried to access it. So what's the recommended topology to support large number of services and get better performance? For now, we have two separate kube-proxy hosts as the publicIP and also running kube-proxy service on each minion.

5. "You really need to use kube-proxy to take full advantage of kubernetes. ", do you mean service choreography here, or something else?

6. If we want to replace kube-proxy with HA-Proxy, do you think we can get better performance? 

Thanks a lot!

**Comment by lavalamp** [comment]
> we have thousands of rules created in iptables on each node that has kube-proxy

@thockin have we done much performance testing with kube-proxy?

@nkwangleiGIT If I understand what you're describing, it sounds like you are building something like AppEngine. kube-proxy's main purpose is to route traffic from pods to services inside the cluster; it also can route traffic via the old PublicIP method.

My recommendations for your setup:
* Make every independent RC/Pod/Service inside its own namespace; this should make things like ENV vars passed into containers scale much better. (assuming that you are not trying to talk to services in other namespaces!)
* I think your approach of having some dedicated front-end machines to be the endpoint for incoming traffic is good. kube-proxy can possibly work for this, but it's not sharded so it isn't going to scale well. You may be better off implementing actual load balancers (via HAproxy, nginx, or cloud-specific load balancers).

Unfortunately, we haven't yet optimized for your use case, so you're probably going to hit some rough edges.

**Comment by nkwangleiGIT** [comment]
Hi lavalamp, thanks for your comments.
1. We already used namespace for each user for each RC/pod/service
2. Yes, we found one weird issue on kube-prxy below:
https://github.com/GoogleCloudPlatform/kubernetes/issues/9793
So for now, we're going to use HAProxy to replace kube-proxy to fix this, please see if any comments for 2nd issue, thanks.

**Comment by smarterclayton** [comment]
We've tested haproxy up to hundreds of thousands of front ends - 1.5 has a number of performance improvements and dynamic acl rules to enable that case.  It depends on how much traffic you want to put through your services.  In our case, we are programming the backend from the service endpoints, but the principle is the same.

**Comment by smarterclayton** [comment]
We probably also want to limit or scope service rules on a node in the future (so that you don't necessarily see all the services in the entire cluster on the same node).  A more in depth design has not been proposed yet

**Comment by thockin** [comment]
I tested just plain iptables and I did not see any measurable slowdown with
tens of thousands of rules.

We have done some measurements of kube-proxy, and we know it has a long
tail.  Plans to fix that are after 1.0

**Comment by smarterclayton** [comment]
We've had rhel 6 instances with a hundred thousand rules performing port mapping (similar but not identical use pattern) and the primary issues were restart and update times, not network.

**Comment by nkwangleiGIT** [comment]
Hi smarterclayton,
For now, w're using HAProxy 1.5.8 in our environment, and it's much quicker than kube-proxy now. And we will get the service and endpoints information periodically to generate frontend and backend into haproxy configuration and then reload haproxy config. It needs around 1 second refresh haproxy.
Do you have any better approach to achieve this? for example, we can listen to kube-apiserver and get the notification and then update haproxy. But I didn't find any document about how to do this using k8s, please let me know if any suggestions, thanks!

**Comment by davidopp** [comment]
> Do you have any better approach to achieve this? for example, we can listen to kube-apiserver
> and get the notification and then update haproxy. But I didn't find any document about how to
> do this using k8s,  please let me know if any suggestions, thanks!

I think what you want is exactly what you described at the beginning of your comment -- "get the service and endpoints information periodically to generate frontend and backend into haproxy configuration and then reload haproxy config." You can do it exactly the way you described -- create a service with a label selector corresponding to a label that you put on the pods you're interested in, and then set something up that watches the service's endpoints and uses the information to continuously reconfigure HAProxy.

**Comment by mikedanese** [comment]
Take a look at https://github.com/mikedanese/k8s-haproxy/ for an example of this, although it hasn't been updated for v1 api, multiport services, or the new client auth requirements. I know there are other project that do this as well. It would be nice to get something in contrib because it's such a common requirement. It would also be great if the current refactoring of the proxy interface would allow us to someday implement an haproxy backend. cc @BenTheElder

**Comment by BenTheElder** [comment]
https://github.com/GoogleCloudPlatform/kubernetes/pull/9210 should hopefully help a lot with performance (issue: https://github.com/GoogleCloudPlatform/kubernetes/issues/3760). A prototype in use was mentioned here:
https://github.com/GoogleCloudPlatform/kubernetes/issues/3760#issuecomment-104823239

@mikedanese I'd be happy to work on something like that once the iptables rewrite is finished; I'm still trying to verify that it works correctly and land the relevant PR's. Vagrant + e2e breaking put me a little behind.

**Comment by davidwalter0** [comment]
@thockin Is there an example and description of an iptables configuration for / replacement of kube-proxy or enablement in kube-proxy.

**Comment by mikedanese** [comment]
@davidwalter0  #3760 has a few proof of concept rules.

**Comment by nkwangleiGIT** [comment]
@mikedanese, thanks, will track the project you recommended below:
https://github.com/mikedanese/k8s-haproxy/
Seems we're also replacing userspace proxy with iptables, I think it'll probably bring much better performance, but not sure if this feature will go to v1, can we confirm this now?

Thanks for all who added comments to this issue, it's helpful for me!

**Comment by davidopp** [comment]
iptables instead of user-space kube-proxy will not be in the 1.0 release

**Comment by erictune** [comment]
I'm closing this to show that no more action is needed.  Feel free to open new issues with other questions.

**Comment by nkwangleiGIT** [comment]
Hi erictun, I'm just going to close this one, and thanks all for the kindly help!
Currently, we're using HAProxy and it's much faster, will open another one if any other issues.
Thanks again!

**Comment by jjfiguep** [comment]
Hi,

First of all, this is my first comment in GitHub and sorry if this is not the correct place.

In previous comments it is discussed the possibility to use Kubernetes and HAProxy. Also it is commented the option to periodically detect new services and update haproxy config file. 
This is just my problem, I´m using NodePort to expose the services to internet but I would like replace the assigned port number to access to the service by something more user friendly and my idea is to use haproxy to do the same (when /foo is requested in the url, haproxy should redirect the traffic to the corresponding port). The problem is that my implementation of Kubernetes is continuously creating and destroying services and the process to update haproxy configuration should be automatic. My doubts below:
1.- Am I focusing the problem from the correct side?
2.- If not, how I can resolve my problem? If yes, there is some project or something to do this? or could someone help me with this?

Thanks a lot!!

**Comment by erictune** [comment]
@jjfiguep Your question is better asked in a new issue.  I've copied it to #12665

**Comment by andrejvanderzee** [comment]
Hi Wang Lei,

We are facing similar issues. Could you please elaborate how you are using haproxy? Are you watching the kube-apiserver endpoints and hot reload the config of haproxy whenever a targeted endpoint changes? Does that scale to 10k-100k pods? 

Thanks!
